vendors/vendor/boards/board/ports/ota_pal_for_aws/ota_pal.c

```diff
@@ -0,0 +1,183 @@
+ * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+ *
+ * http://aws.amazon.com/freertos
+ * http://www.FreeRTOS.org
+ */
+
+/* C Runtime includes. */
+#include <stdio.h>
+#include <stdlib.h>
+
+/* FreeRTOS include. */
+#include "FreeRTOS.h"
+
+/* OTA library includes. */
+#include "ota.h"
+#include "ota_private.h"
+
+/* OTA PAL Port include. */
+#include "ota_pal.h"
+
+/* OTA PAL test config file include. */
+#include "ota_config.h"
+
+/* Specify the OTA signature algorithm we support on this platform. */
+const char OTA_JsonFileSignatureKey[ OTA_FILE_SIG_KEY_STR_MAX_LENGTH ] = "sig-sha256-ecdsa"; /* FIX ME. */
+
+
+/* The static functions below (otaPAL_CheckFileSignature and otaPAL_ReadAndAssumeCertificate)
+ * are optionally implemented. If these functions are implemented then please set the following
+ * macros in aws_test_ota_config.h to 1:
+ *   otatestpalCHECK_FILE_SIGNATURE_SUPPORTED
+ *   otatestpalREAD_AND_ASSUME_CERTIFICATE_SUPPORTED
+ */
+
+/**
+ * @brief Verify the signature of the specified file.
+ *
+ * This function should be implemented if signature verification is not offloaded to non-volatile
+ * memory io functions.
+ *
+ * This function is expected to be called from otaPal_CloseFile().
+ *
+ * @param[in] pFileContext OTA file context information.
+ *
+ * @return The OtaPalStatus_t error code combined with the MCU specific error code. See
+ *         ota_platform_interface.h for OTA major error codes and your specific PAL implementation
+ *         for the sub error code.
+ *
+ * Major error codes returned are:
+ *   OtaPalSuccess: if the signature verification passes.
+ *   OtaPalSignatureCheckFailed: if the signature verification fails.
+ *   OtaPalBadSignerCert: if the signature verification certificate cannot be read.
+ */
+static OtaPalStatus_t otaPal_CheckFileSignature( OtaFileContext_t * const pFileContext );
+
+/**
+ * @brief Read the specified signer certificate from the filesystem into a local buffer.
+ *
+ * The allocated memory returned becomes the property of the caller who is responsible for freeing it.
+ *
+ * This function is called from otaPAL_CheckFileSignature(). It should be implemented if signature
+ * verification is not offloaded to non-volatile memory io function.
+ *
+ * @param[in] pucCertName The file path of the certificate file.
+ * @param[out] ulSignerCertSize The size of the certificate file read.
+ *
+ * @return A pointer to the signer certificate in the file system. NULL if the certificate cannot be read.
+ * This returned pointer is the responsibility of the caller; if the memory is allocated the caller must free it.
+ */
+static uint8_t * otaPal_ReadAndAssumeCertificate( const uint8_t * const pucCertName,
+                                                  uint32_t * const ulSignerCertSize );
+
+
+/*-----------------------------------------------------------*/
+
+OtaPalStatus_t otaPal_Abort( OtaFileContext_t * const pFileContext )
+{
+    /* FIX ME. */
```

Comment:
I like the idea of marking where the user needs to fill in the PAL functions. 
It would be nicer if they used the TODO keyword.  This will be caught by doxygen and IDE's to make a work list.
It would also be nice to have a rough outline of what is needed in each function so the user can fill in the details but not have to think about what is the purpose of the function.

Reply:
Updated this to be TODO comments.

These are  dependent on the hardware or use of file system so I'm not sure how detailed of an outline I could provide. I tried it out for a couple of these and I don't think it provided much more information than just the function description block.

Reply:
I will change my mind to a TODO that just provides a task i.e. 
void foo(void)
{
 /* TODO implement foo */
}

Reply:
Nice idea, done.